File: dipgos-backend/app/models/ccc.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional, List, Dict, Any

from pydantic import BaseModel, Field


class CCCSelection(BaseModel):
    tenant_id: str = Field(default="default", description="Tenant identifier")
    project_id: str
    contract_id: Optional[str] = None
    sow_id: Optional[str] = None
    process_id: Optional[str] = None


class MapMarker(BaseModel):
    id: str
    type: Literal["contract", "sow", "process"]
    name: str
    lat: float
    lon: float
    status: Literal["on-track", "monitoring", "risk"]
    percent_complete: float = Field(ge=0.0, le=100.0)
    spi: Optional[float] = None
    cpi: Optional[float] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)


class WipDial(BaseModel):
    id: str
    level: Literal["project", "contract", "sow", "process"]
    code: Optional[str] = None
    name: str
    percent_complete: float = Field(ge=0.0, le=100.0)
    ev: Optional[float] = None
    pv: Optional[float] = None
    ac: Optional[float] = None
    spi: Optional[float] = None
    cpi: Optional[float] = None


class CccSummary(BaseModel):
    selection: CCCSelection
    map: List[MapMarker]
    wip: List[WipDial]
    as_of: datetime


class PhysicalWorksCard(BaseModel):
    actual_percent: Optional[float] = None
    planned_percent: Optional[float] = None
    variance_percent: Optional[float] = None
    trend_actual: List[float] = Field(default_factory=list)
    trend_planned: List[float] = Field(default_factory=list)
    notes: List[str] = Field(default_factory=list)


class WorkInProgressCategory(BaseModel):
  name: str
  count: int = 0
  planned_percent: Optional[float] = None
  actual_percent: Optional[float] = None
  variance_percent: Optional[float] = None


class WorkInProgressCard(BaseModel):
    categories: List[WorkInProgressCategory] = Field(default_factory=list)
    notes: List[str] = Field(default_factory=list)


class WorkOutputItem(BaseModel):
    name: str
    planned_percent: Optional[float] = None
    actual_percent: Optional[float] = None
    variance_percent: Optional[float] = None


class WorkOutputCard(BaseModel):
    items: List[WorkOutputItem] = Field(default_factory=list)
    notes: List[str] = Field(default_factory=list)


class QualitySummaryCard(BaseModel):
    ncr_open: int = 0
    ncr_closed: int = 0
    qaor_open: int = 0
    qaor_closed: int = 0
    quality_conformance: Optional[float] = None


class PerformanceSnapshotCard(BaseModel):
    spi: Optional[float] = None
    cpi: Optional[float] = None
    ev: Optional[float] = None
    pv: Optional[float] = None
    ac: Optional[float] = None
    burn_rate_days: Optional[float] = None
    runway_days: Optional[float] = None
    cash_flow: Optional[float] = None
    trend_spi: List[float] = Field(default_factory=list)
    trend_cpi: List[float] = Field(default_factory=list)
    notes: List[str] = Field(default_factory=list)


class RightPanelKpiPayload(BaseModel):
    selection: CCCSelection
    as_of: datetime
    physical: PhysicalWorksCard
    work_in_progress: WorkInProgressCard
    work_output: WorkOutputCard
    performance: PerformanceSnapshotCard
    preparatory: WorkOutputCard
    quality_summary: Optional[QualitySummaryCard] = None

File: src/types/ccc.ts
export type CccSelection = {
  tenant_id: string
  project_id: string
  contract_id?: string | null
  sow_id?: string | null
  process_id?: string | null
}

export type MapMarker = {
  id: string
  type: 'contract' | 'sow' | 'process'
  name: string
  lat: number
  lon: number
  status: 'on-track' | 'monitoring' | 'risk'
  percent_complete: number
  spi?: number | null
  cpi?: number | null
  metadata?: Record<string, unknown>
}

export type WipDial = {
  id: string
  level: 'project' | 'contract' | 'sow' | 'process'
  code?: string | null
  name: string
  percent_complete: number
  ev?: number | null
  pv?: number | null
  ac?: number | null
  spi?: number | null
  cpi?: number | null
}

export type CccSummary = {
  selection: CccSelection
  map: MapMarker[]
  wip: WipDial[]
  as_of: string
}

export type PhysicalWorksCard = {
  actual_percent?: number | null
  planned_percent?: number | null
  variance_percent?: number | null
  trend_actual: number[]
  trend_planned: number[]
  notes?: string[]
}

export type WorkInProgressCategory = {
  name: string
  count: number
  planned_percent?: number | null
  actual_percent?: number | null
  variance_percent?: number | null
}

export type WorkInProgressCard = {
  categories: WorkInProgressCategory[]
  notes?: string[]
}

export type WorkOutputItem = {
  name: string
  planned_percent?: number | null
  actual_percent?: number | null
  variance_percent?: number | null
}

export type WorkOutputCard = {
  items: WorkOutputItem[]
  notes?: string[]
}

export type QualitySummaryCard = {
  ncr_open: number
  ncr_closed: number
  qaor_open: number
  qaor_closed: number
  quality_conformance?: number | null
}

export type PerformanceSnapshotCard = {
  spi?: number | null
  cpi?: number | null
  ev?: number | null
  pv?: number | null
  ac?: number | null
  burn_rate_days?: number | null
  runway_days?: number | null
  cash_flow?: number | null
  trend_spi: number[]
  trend_cpi: number[]
  notes?: string[]
}

export type RightPanelKpiPayload = {
  selection: CccSelection
  as_of: string
  physical: PhysicalWorksCard
  work_in_progress: WorkInProgressCard
  work_output: WorkOutputCard
  performance: PerformanceSnapshotCard
  preparatory: WorkOutputCard
  quality_summary?: QualitySummaryCard
}
